pyface/data_view/abstract_data_model.py

```diff
@@ -0,0 +1,272 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Provides an AbstractDataModel ABC for Pyface data models.
+
+This module provides an ABC for all data view data models.  This specifies
+the API that the data view widgets expect, and which the underlying
+data is adapted to by the concrete implementations.  Data models are intended
+to be toolkit-independent, and be able to adapt any approximately tabular or
+nested data structure to what the data view system expects.
+"""
+from abc import abstractmethod
+
+from traits.api import ABCHasStrictTraits, Event, Instance
+
+from .index_manager import AbstractIndexManager
+
+
+class DataViewSetError(ValueError):
+    """ An exception raised when setting a value fails. """
+    pass
+
+
+class AbstractDataModel(ABCHasStrictTraits):
+    """ Abstract base class for Pyface data models.
+
+    The data model API is intended to provide a common API for hierarchical
+    and tabular data.  This class is concerned with the structure, type and
+    values provided by the data, but not with how the data is presented.
+
+    Row and column indices are represented by sequences (usually lists) of
+    integers, specifying the index at each level of the hierarchy.  The root
+    row and column are represented by empty lists.
+
+    Subclasses need to implement the ``get_column_count``,
+    ``can_have_children`` and ``get_row_count`` methods to return the number
+    of columns in a particular row, as well as the hierarchical structure of
+    the rows.  Appropriate observers should be set up on the underlaying data
+    so that the ``structure_changed`` event is fired when the values returned
+    by these methods would change.
+
+    Subclasses also have to implement the ``get_value`` and ``get_value_type``
+    methods.  These expect a row and column index, with root values treated
+    specially: the root row corresponds to the values which will be displayed
+    in the column headers of the view, and the root column corresponds to the
+    values which will be displayed in the row headers of the view.
+    The ``get_value`` returns an arbitrary Python object corresponding to the
+    cell being viewed, and the ``get_value_type`` should return an instance of
+    an ``AbstractValueType`` that adapts the raw value to the data channels
+    that the data view expects (eg. text, color, icons, editable value, etc.).
+    Implementations should ensure that the ``values_changed`` event fires
+    whenever the data, or the way the data is presented, is updated.
```

Comment:
Nit: This sentence may be better moved to the next paragraph which has to do with changing values via ``set_value``. The above paragraph concerns "getting" the value.